### ISSUE TITLE
'headers already sent' from ArtifactResolutionService

### ISSUE THREAD

**Issue by timforsyth** [simplesamlphp/simplesamlphp#413]
**edited as I have now tried this with a browser **

Hi,

I've recently had an issue with Artifact Binding.  

After much hair tearing, I've tracked the issue down to a potential bug in SimpleSAML_Logger

Here's the response from ArtifactResolutionService.php:

```
<?xml version="1.0" encoding="UTF-8"?><SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/"><SOAP-ENV:Body><samlp:ArtifactResponse 

......

</samlp:ArtifactResponse></SOAP-ENV:Body></SOAP-ENV:Envelope>
Warning: Cannot modify header information - headers already sent by (output started at /simplesamlphp/vendor/simplesamlphp/saml2/src/SAML2/SOAP.php:29) in /simplesamlphp/lib/SimpleSAML/Utils/HTTP.php on line 984

Fatal error: Exception thrown without a stack frame in Unknown on line 0
```

It took a while to track down but the issue is with SimpleSAML_Logger::flush

::flush is called from a shutdown function registered in SimpleSAML_Logger::defer

If no session is present, the flush function tries to create a session and set a cookie (SimpleSAML_Session::getSessionFromRequest)

As this is happening during shutdown, the output buffer has already been written to and you get this error.

This happened when I was writing an integration test for our SAML server and after I upgraded to V1.14 from V1.10.

**edit**
Now I have tried this with a real browser: 

The tricky thing here is that there is a session identifier, in a cookie stored in the SAML IDP domain, however during the SAMLArt redirect to the SP, the SP is not given this session identifier, which seems to cause the header already sent error.

Am I missing something obvious, or is this a bug ?
 
I should add I am running php 5.3.3 and I just found out that I can make this issue go away if I set the output buffer size in PHP to > 9K (this size of the SAML Artifact Resolve response document). This workaround doesn't sit well with me.

**Comment by jaimeperez** [comment]
Hi @timforsyth!

Thanks for the report. This is indeed a bug, which can be easily replicated with the following script:

```php
<?php

include(PATH_TO_SSP.'www/_include.php');
header('Content-Type: text/plain');
echo "Some text\n";
flush();
SimpleSAML\Logger::critical('Critical error');
```

On the other hand, even though there is a bug and SSP should not trigger an exception in this case, there's also a misconfiguration on your IdP, which should never print out errors and warnings to standard output. You might have the `display_errors` configuration option in your *php.ini* set to `1`, and that's why your artifact responses are broken. Turning that off should fix the problem for you, at least momentarily meanwhile I fix the original bug.

**Comment by timforsyth** [comment]
Great, thanks Jamie!

By the way, display_errors is set to Off in my container, which is odd but no worries, so I'll look out for a future release.

thanks,
Tim.

**Comment by jaimeperez** [comment]
Hi again!

I've just committed a fix for this. Could you please update your SimpleSAMLphp installation to the commit with the fix, and tell me if it works for you? I have verified that indeed it solves the issue when reproduced with the simple script I used in my previous comment.

If everything works fine, I'll cherry pick the set of commits fixing the issue from the 1.14 branch, and prepare a new release with the fix and some other stuff, probably.

**Comment by timforsyth** [comment]
Hi!

I've installed from the master branch and can confirm this has fixed the issue.

Thanks for the quick turn around, I'll keep an eye out for a new release.

cheers,
Tim.

**Comment by jaimeperez** [comment]
Hi @timforsyth!

Thanks for the feedback. Just a couple of comments on this:

- I've been replicating your use case, and I can confirm even with the previous bug, I didn't get the error as part of the artifact response. It must be some misconfiguration of either SimpleSAMLphp or the web server, that made your IdP append that error to the actual response.

- I've also committed some new fixes related to this issue and #356, which was related. Both should be completely solved now, and even in case we ever have another similar issue, it shouldn't affect your use case as it did before.
